# vcf2msa
## Python class for one-hot encoding variants from vcf files, then creating a multiple sequence alignment (MSA) as a PHYLIP file

N.B. Designed for vcf files from single calling, not joint

N.B. Only supports uniploid genomes for now, with the intended use being for bacterial genomes

/// --------------------------------------- //

### CLASS INIT

onehot_variants class is initialized with no parameters.

/// --------------------------------------- ///

### FUNCTIONS

onehot_variants class has the following functions:

/// ---------- ///

process_vcfs(vcf_files, vcf_ids)

	Processes a list of vcf files and one-hot encodes variants.

	Attributes:

	vcf_files
					List of paths to vcf files

	vcf_ids
					List of IDs for vcf files

/// ---------- ///

save_data(file_prefix='./vcf_data', full_matrix=False)

	Saves to file the one-hot encoded variants and relevant informations about variants and samples.
	Three files are generated:
	- *.txt = vcf file IDs;
	- *.tsv = table with info about variants;
	- *.npz = sparse matrix file of one-hot enocoded sample variants; (Only if full_matrix = False)
	- *.mtx = full matrix file of one-hot enocoded sample variants; (Only if full_matrix = True)

	Attributes:

	file_prefix
					Prefix of output files
					Default = './vcf_data'

	full_matrix
					Whether the matrix of one-hot encoded values should be output as a sparse matrix (*.npz) or full matrix (*.mtx)
					Default = False

/// ---------- ///

load_data(file_prefix='./vcf_data', full_matrix=False)

	Loads files generated by the save_data function.

	Attributes:

	file_prefix
					Prefix of input files
					Default = './vcf_data'

	full_matrix
					Whether the matrix of one-hot encoded values is a sparse matrix (*.npz) or full matrix (*.mtx)
					Default = False

/// ---------- ///

build_phylip_msa(var_only=True, snp_only=False, out_name='msa.phy')

	Parses the variant info and creates a multiple sequence alignment in PHYLIP format.

	Attributes:

	var_only
					Set to True to output a MSA with only variant sites
					Default = True

	snp_only
					Set to True to use only SNPs for generating the MSA, removing insertions and indels
					Default = False

	out_name
					Output name for the PHYLIP MSA file
					Default = 'msa.phy'

/// ---------- ///

build_reference_genome(var_info, var_only=True, snp_only=False)

	Static method for creating reference chromosomes based on variants reference information

	Attributes:

	var_info
					Pandas data frame with variants information (from onehot_variants.variants_info)

	var_only
					Set to True to remove constant sites
					Default = True

	snp_only
					Set to True to use only SNPs for generating the MSA, removing insertions and indels
					Default = False

/// --------------------------------------- ///

### DEPENDENCIES:

Python 3.9.16+ &

	numpy
	pandas
	scipy.sparse

/// --------------------------------------- ///

### NOTES:

- MSA can be generate using all sites or variant sites only.

- MSA can be generated using all variants or SNPs only. This is useful as input to (e.g.) IQ-TREE2 with model option +ASC.

/// --------------------------------------- ///

### USAGE EXAMPLE:

```
	import onehot_encode_variants

	from os import listdir

	vcf_files = [file for file in listdir() if file.endswith('.vcf.gz')]
	vcf_ids = [f'sample_{i}' for i in range(len(vcf_files))]

	# One-hot encode variants
	encodings = onehot_encode_variants.onehot_variants()
	try:
	    
	    encodings.load_data(file_prefix='./variant_data')

	except:
	    
	    encodings.process_vcfs(vcf_files, vcf_ids)
	    encodings.save_data(file_prefix='./variant_data')

	# Create MSA
	encodings.build_phylip_msa(var_only=True, snp_only=False, out_name='msa_complete.phy')

	# Create MSA with SNPs only
	encodings.build_phylip_msa(var_only=True, snp_only=True, out_name='msa_snp_only.phy')
```
